examples/plot_gp_function.py

```diff
@@ -0,0 +1,270 @@
+        utility_function[acq[count]] = UtilityFunction(kind=acq[count], kappa = params[acq[count]]['kappa'], xi = params[acq[count]]['xi'])
+        utility[acq[count]] = utility_function[acq[count]].utility(x, optimizer._gp, optimizer._space.target.max())
+        count = count+1
+'''
+    
+    x_obs = []
+    y_obs = []
+    mu=[]
+    sigma=[]
+    utility=[]
+    
+    acq = ['ucb', 'ei', 'poi']
+    for i, optimizer in enumerate(optimizers): # enumerate applicata a iterable -> sia valore che indice
+        x_obs.append(np.array([[res["params"]["x"]] for res in optimizer.res]))
+        y_obs.append(np.array([res["target"] for res in optimizer.res]))
+        mu_temp, sigma_temp = posterior(optimizer, x_obs[i], y_obs[i], x)
+        mu.append(mu_temp)
+        sigma.append(sigma_temp)
+        utility_function = UtilityFunction(kind=acq[i], kappa = params[acq[i]]['kappa'], xi = params[acq[i]]['xi'])
+        utility.append(utility_function.utility(x, optimizer._gp, optimizer._space.target.max()))
+        
+
+    
+    axis.plot(x, y, linewidth=3, label='Target')
+
+    for i, mean in enumerate(mu):
+        axis.plot(x, mean, '--', color= colmap(i), label='Prediction {}'.format(acq[i]) )
+
+    
+    fig.suptitle(
+        'Utility Functions After {} Steps'.format(steps-n_init),
+        fontdict={'size':50}
+    )
+    
+    axis.set_xlim((min(x), max(x)))
+    axis.set_ylim((None, None))
+    axis.set_ylabel('f(x)', fontdict={'size':20})
+    axis.set_xlabel('x', fontdict={'size':20})
+ 
+
+    for i, util in enumerate(utility):
+        af.plot(x, normalization(util), label=acq[i], color= colmap(i))
+        af.plot(x[np.argmax(util)], np.max(normalization(util)), '*', markersize=15, markerfacecolor=colmap(i), markeredgecolor='k', markeredgewidth=1)
+
+
+    af.set_xlim((min(x), max(x)))
+    af.set_ylim((-0.5,1.5))
+    af.set_ylabel('Utility', fontdict={'size':20})
+    af.set_xlabel('x', fontdict={'size':20})
+    
+
+    
+    axis.legend(loc=2, bbox_to_anchor=(1.01, 1), borderaxespad=0.)
+    af.legend(loc=2, bbox_to_anchor=(1.01, 1), borderaxespad=0.)
+    
+    plt.show()
+    
+    
+    
+def plot_convergence(optimizers, x, target, params, it=20):
+    x = x.reshape(-1,1)
+    y = target(x)
+    
+    tar={}
+    points={}
+
+    utility_function={}
+    for acq, optimizer in optimizers.items():  # acq è la stringa-chiave, optimizer è l'oggetto
+        utility_function[acq] = UtilityFunction(kind=acq, kappa = params[acq]['kappa'], xi = params[acq]['xi'])
+        
+
+    for acq, optimizer in optimizers.items():
+            points[acq]=np.zeros(it)
+            tar[acq]= np.zeros(it)
+            
+    for i in range(it):
+        for acq, optimizer in optimizers.items():
+            points[acq][i]=optimizer.suggest(utility_function[acq])['x']
+            tar[acq][i]= target(points[acq][i])
+            optimizer.register(params=points[acq][i], target=tar[acq][i])
```

Comment:
Questi tre `for` potrebbero essere inglobati in un unico ciclo. Questo anche ipotizzando che possiate scambiare l'ordine delle righe 135 e 136, ma mi sembra di sì